[SOL-1998] Implement Show Answer button

arbrandes opened the pull request:
## Description

In assessment mode, when the learner has run out of attempts, enable a "Show Answer" button. Clicking it causes all items to move to a correct placement.

## Dependencies

*None*

## JIRA

- [SOL-1998](https://openedx.atlassian.net/browse/SOL-1998)

## Sandbox URLs

- [Assessment mode](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/): a DnDv2 instance in assessment mode, with one maximum attempt.
- [Standard mode](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2_standard/): a DnDv2 instance in standard mode.

## Latest commit

- adcbf1c *(updated 2016-09-17)*

## Testing instructions

### Assessment mode

1. Go to the [assessment mode DnDv2 instance](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2/) on the LMS, and log in as `staff@example.com`.
2. Click on `Staff Debug Info`, `Delete Student State`, and refresh the page.
3. Notice that there is a "Show Answer" button, because this is in assessment mode.
4. Notice that the "Show Answer" button is disabled, because this is your first attempt.  
5. Pick any item and drop it onto any zone.
6. Click submit, which will make you run out of attempts (there's only one).
7. Notice that the "Show Answer" button is now enabled.
8. Click the "Show Answer" button, see that all items are placed correctly (and decoy items are still in the bank), and check that you can't drag any of them.
9. Refresh the page, and see that your final attempt is displayed.
10. Click "Show Answer" again.  The same thing should happen again.

### Standard mode

1. Go to the [standard mode DnDv2 instance](http://dndv2-sol-1998.opencraft.hosting/courses/course-v1:test+tt101+run/courseware/chapter/drag_and_drop_v2_standard/) on the LMS, and log in as `staff@example.com`.
2. Notice that there is no "Show Answer" button, because this is in standard mode.

arbrandes commented:
@bradenmacdonald, two favors, when you have a couple of minutes:

* Can I get an `dndv2-sol1998` sandbox for this?
* Can you please update the https://github.com/open-craft/xblock-drag-and-drop-v2 to edx-solutions master?

bradenmacdonald commented:
@arbrandes  For this task, @itsjeyd will be your reviewer. @itsjeyd can you please help Adolfo with those two items ^ ? (And when this is ready for upstream, Tim please open the upstream PR yourself so that it doesn't create an OSPR ticket - and then copy your +1 over)

itsjeyd commented:
Thanks @arbrandes for the PR!

* The sandbox is currently provisioning. I'll post Studio and LMS links here when it's done.
* I updated `open-craft:master` with the latest changes from `edx-solutions:master`. Was surprised to see that the diff for this PR is still showing the commits that are now present in `open-craft:master`. Based on [this](http://stackoverflow.com/a/38016309/1199226), you'll need to rebase to reduce the diff to the relevant commits. I'll have a look at the code when you're done with that; let me know.

arbrandes commented:
@itsjeyd, thanks for the sandbox!

And yes, I'm aware that PRs don't track target branch changes.  Unfortunately, just rebasing doesn't work if the target branch is updated with the same base commits as the PR: git doesn't detect a change, and thus doesn't trigger a refresh.

However, I did mean to improve the README a little.  After a push force, it seems to have done the trick. ;)

itsjeyd commented:
@arbrandes The sandbox is up-and-running now.

Studio: http://studio-dndv2-sol-1998.opencraft.hosting/
LMS: http://dndv2-sol-1998.opencraft.hosting/

Note that the DNS is still updating, so it might take a little while longer for the URLs above to become active.

arbrandes commented:
@itsjeyd, thank you!  I've deployed a test course.  The testing instructions above should now work.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -445,6 +458,21 @@ def _validate_do_attempt(self):
 
+    def _validate_show_answer(self):
+        """
+        Validates if `show_answer` handler should be executed
```
@arbrandes Nit: Add period to the end of this sentence. Also, it would make sense to replicate info about possible exceptions from L392-L394 here.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -445,6 +458,21 @@ def _validate_do_attempt(self):
                 self.i18n_service.gettext("Max number of attempts reached")
             )
 
+    def _validate_show_answer(self):
```
@arbrandes Is this called from multiple places? If not, I think it would be perfectly fine to move the checks to the `show_answer` handler and remove this method.

itsjeyd commented on drag_and_drop_v2/drag_and_drop_v2.py:
```diff
@@ -672,6 +700,36 @@ def _get_user_state(self):
+                if zone is not None and zone != 'none':
+                    zones.append(zone)
+
+            if len(zones):
```
@arbrandes You can just do `if zones:` here I think.

arbrandes replied:
Not really.  In the new version, `[]` is a valid zone definition, but it evaluates to false.  We only want to go into that block if we detect an old version of data, where `zones === None`.

itsjeyd replied:
@arbrandes Hm. I'm probably missing something/not interpreting your comment correctly, but for the case where `zones` is `[]`, both `if zones:` and `if len(zones):` would evaluate to false, right? And once execution reaches this line (L719), `zones` can not be `None` (the `if` block that starts in L713 above takes care of that). So I don't see why it's necessary to do `if len(zones):` here yet. Please elaborate? :)

arbrandes replied:
Ok, it was I who didn't interpret your comment correctly.  You're right, on line 719 we can certainly do without the `len()`.

arbrandes replied:
Addressed in 89839ec, sorry about the confusion.

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -212,9 +212,14 @@ function DragAndDropTemplates(configuration) {
+
         if (zone.align !== 'none') {
           item_wrapper += '.item-align.item-align-' + zone.align;
-          //items_in_zone = $.grep(ctx.items, is_item_in_zone);
```
@arbrandes This looks like it might still be needed? Not the comment, the actual code :)

arbrandes replied:
I'm not a fan of leaving code commented out.  It's why we have SCM, after all.  Should I really leave it in?

itsjeyd replied:
@arbrandes I wasn't talking about leaving the comment in, I was wondering if the code itself was still needed. The `else` branch below modifies `items_in_zone`, but the `if` branch currently doesn't. That asymmetry looked suspicious to me.

So, another way of putting the question would be: Why does the `if` branch get away with not modifying `items_in_zone`?

arbrandes replied:
To answer your question, because it's handled above, on line 200, and the author of this changeset probably forgot the commented line in.

On the merits of the implemented logic, the point of this `if/else` block is to avoid rendering any unaligned items in the zone template: these are rendered elsewhere.

itsjeyd replied:
@arbrandes 

> because it's handled above, on line 200, and the author of this changeset probably forgot the commented line in.

Thanks, that makes sense. I hadn't expanded the diff to see more context.

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -356,14 +361,14 @@ function DragAndDropTemplates(configuration) {
+                    {disabled: disabled || spinner || false, attributes: {tabindex: 0}},
                     [
-                        h("span.btn-icon.fa."+iconClass, {attributes: {"aria-hidden": true}}, []),
+                        h("span.btn-icon.fa." + (spinner ? 'fa-spin.fa-spinner' : iconClass), {attributes: {"aria-hidden": true}}, []),
```
@arbrandes This would be easier to read like this:

```js
        if (spinner) {
            iconClass = 'fa-spin.fa-spinner';
        }
        return (
            // ...
                        h(
                            "span.btn-icon.fa." + iconClass,
                            {attributes: {"aria-hidden": true}}, 
                            []
                        ),
                        // ...
```

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -376,6 +381,7 @@ function DragAndDropTemplates(configuration) {
             h("section.action-toolbar-item.sidebar-buttons", {}, [
                 sidebarButtonTemplate("keyboard-help-button", "fa-question", gettext('Keyboard Help')),
                 sidebarButtonTemplate("reset-button", "fa-refresh", gettext('Reset'), ctx.disable_reset_button),
+                (ctx.show_show_answer ? sidebarButtonTemplate("show-answer-button", "fa-warning", gettext('Show Answer'), ctx.disable_show_answer_button, ctx.show_answer_spinner) : null),
```
@arbrandes This line is quite long and a little hard to read. I'd probably spell the logic out before the call to `h`:

```js
var showAnswerButton = null;
if (ctx.show_show_answer) {
    showAnswerButton = sidebarButtonTemplate(
        "show-answer-button",
        "fa-warning",
        gettext('Show Answer'),
        ctx.disable_show_answer_button,
        ctx.show_answer_spinner
    );
}
```

... and then just reference the `showAnswerButton` variable here.

itsjeyd commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -1147,8 +1182,11 @@ function DragAndDropBlock(runtime, element, configuration) {
             disable_submit_button: !canSubmitAttempt(),
-            submit_spinner: state.submit_spinner
+            submit_spinner: state.submit_spinner,
+            show_answer: state.show_answer,
```
@arbrandes `state.show_answer` represents information about whether the board is currently displaying the correct answer, right? If so, I think it would make sense to rename `show_answer` to `showing_answer`. (`show_answer` can also be interpreted as "allowed to show answer", so there's some potential for confusion.)

arbrandes replied:
Agreed.  It was a bit confusing, indeed.

itsjeyd commented on tests/integration/test_interaction.py:
```diff
@@ -207,12 +207,17 @@ def assert_placed_item(self, item_value, zone_title, assessment_mode=False):
-            self.assertEqual(item_description.text, 'Correctly placed in: {}'.format(zone_title))
+            description = 'Correctly placed in: {}'
+
+        if isinstance(zone_title, list):
```
@arbrandes Under what circumstances would `zone_title` be a `list`? It might make sense to add a comment explaining that above this line.

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -148,6 +158,42 @@ def test_max_attempts_reached_submit_and_reset_disabled(self):
+        """
+        self.place_item(0, TOP_ZONE_ID)
+        show_answer_button = self._get_show_answer_button()
+        self.assertTrue(show_answer_button.is_displayed())
```
@arbrandes The "Show Answer" should probably be displayed right away, without having to place an item first? If so, it would make sense to do the check first (i.e., to move the `self.place_item(0, TOP_ZONE_ID)` action from L166 below the `self.assertTrue(show_answer_button.is_displayed())` check).

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -148,6 +158,42 @@ def test_max_attempts_reached_submit_and_reset_disabled(self):
+            self.assertEqual(show_answer_button.get_attribute('disabled'), 'true')
+            self.click_submit()
+
+        self.assertEqual(show_answer_button.get_attribute('disabled'), None)
```
@arbrandes Nit: You can use `assertIsNone` here.

itsjeyd commented:
@arbrandes I reviewed the code and left some comments. You also have a quality failure that you'll need to fix. I'll do manual testing on the sandbox (thanks for the nice instructions!) once you're done addressing my feedback, and the sandbox is up-to-date with the latest code.

arbrandes commented on drag_and_drop_v2/public/js/drag_and_drop.js:
```diff
@@ -389,7 +395,7 @@ function DragAndDropTemplates(configuration) {
         var items_placed = $.grep(ctx.items, is_item_placed);
         var items_in_bank = $.grep(ctx.items, is_item_placed, true);
-        var is_item_placed_unaligned = function(i) { return i.zone_align === 'none'; };
+        var is_item_placed_unaligned = function(i) { return i.zone_align === 'none' && !ctx.show_answer; };
```
You're right.  Regardless of the other PR, I'll make the logic clearer.

arbrandes commented:
@itsjeyd, I addressed your comments on bcef5e1, and the sandbox has been updated accordingly.

When you give the ok, I'll squash the fixup(s) for the upstream PR.

itsjeyd commented on tests/integration/test_interaction.py:
```diff
@@ -207,12 +207,20 @@ def assert_placed_item(self, item_value, zone_title, assessment_mode=False):
-            self.assertEqual(item_description.text, 'Correctly placed in: {}'.format(zone_title))
+            description = 'Correctly placed in: {}'
+
+        # Items with multiple possile drop zones could be placed correctly in
```
@arbrandes Typo: "possile"

itsjeyd commented:
Thanks for the updates, @arbrandes! They look good. I'll test the changes manually tomorrow. If you could address the last remaining questions/comments in the meantime, that would be great.

itsjeyd commented:
@arbrandes Just a heads-up that https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/93 hasn't merged yet. It might still happen later on today, but I'm not sure. Let's move forward with the pending rebase so I can test the feature tomorrow morning.

itsjeyd commented:
@arbrandes https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/93 got merged and I updated `open-craft:master` with the new commits. Aside from the changes introduced here, this was the last major feature to land for DnDv2.1. So, once you integrate your changes with the changes from `open-craft:master`, no more complex rebases will come up.

Please go ahead with the following steps, and ping me when you're done so I can have a final look at the code and test the feature on the sandbox:

* Fix typo mentioned [here](https://github.com/open-craft/xblock-drag-and-drop-v2/pull/10#discussion-diff-77014105).
* Squash commits
* Rebase PR on `open-craft:master` and test.
* Update sandbox.

arbrandes commented:
@itsjeyd, all set!  typo fixed, commits squashed, rebased, sandbox updated.

You'll notice I made a few additional changes to the JS that don't relate directly to the title of the PR, but are meant to futher reflect the changes on de5ac0c (i.e., zones can't be unaligned anymore).

arbrandes commented:
Regarding test failures:  the only one I can reproduce is the following, but it happens on master as well.  Looks like a flaky test:

```
ERROR: test_keyboard_help (tests.integration.test_interaction.MultipleBlocksDataInteraction)
(...)
  File "/home/arbrandes/src/edx/xblock-drag-and-drop-v2/tests/integration/test_interaction.py", line 437, in test_keyboard_help
    self.interact_with_keyboard_help(scroll_down=900)
(...)
WebDriverException: Message: Element is not clickable at point (675, 27.866668701171875). Other element would receive the click: <h1 class="title"></h1>
```

itsjeyd commented on tests/integration/test_base.py:
```diff
@@ -384,12 +387,20 @@ def assert_placed_item(self, item_value, zone_title, assessment_mode=False):
+
+        # Items with multiple possible drop zones could be placed correctly in
+        # one of several.  If zone_title is a list, check that it's in any one
+        # of the given zones.
```
@arbrandes 

> ... check that it's in any one of the given zones.

Sorry for not noticing this earlier, but this part of the sentence is ambiguous: It's not clear what "it" is referring to. How about this:

> An item with multiple drop zones could be located in any one of these zones. In that case, `zone_title` will be a list, and we need to check whether the zone info in the description of the item matches any of the zones in that list.

itsjeyd commented on drag_and_drop_v2/translations/en/LC_MESSAGES/text.po:
```diff
@@ -446,6 +454,10 @@ msgid "Reset"
+msgid "Show Answer"
+msgstr ""
+
+#: public/js/drag_and_drop.js
```
It's great to see you updated the translation file for `en`. Could you please update  `eo` file as well? See [this](https://github.com/edx/i18n-tools/blob/master/i18n/dummy.py#L103) for one possible method of generating dummy translations.

itsjeyd commented:
Thanks @arbrandes for the updates! 

I made another pass over the code and only had a couple more comments.

> Regarding test failures: the only one I can reproduce is the following, but it happens on master as well. Looks like a flaky test: ...

How reliably can you reproduce it in your local dev env? If the test fails every time you run it, it should be fixed in the context of this PR.

---

I also tested the feature on the sandbox. Thanks for the nice test instructions! The main functionality (showing correct solution after clicking "Show Answer") is working as expected - nice! :) Here are my notes:

* No way to get back to the submitted solution: When I refresh the page after clicking "Show Answer", incorrectly-placed items are back in the bank; only correctly placed items stay on the board. General feedback displayed below the board changes accordingly. For example:
  * When placing two items correctly, and misplacing the other three items (including the decoy item), general feedback will say "Misplaced 3 items." after clicking "Submit" (as expected). When refreshing the page (or clicking "Show Answer" and then refreshing the page), the message changes to "Did not place 2 required items.
  * When placing a single item incorrectly and leaving the remaining items in the bank, feedback will include "Misplaced 1 item." and "Did not place 3 required items." messages. When clicking "Show Answer", these messages disappear, and I'm left with just "Good work! You have completed this drag and drop problem." and "Final attempt was used, highest score is 0.2".

* After clicking "Show Answer", the button remains enabled. Clicking it again produces the same result each time, so I think it would make sense to disable it once the board has been updated to display the correct solution.

* ~~If the solution submitted via the last remaining attempt is correct, it might make sense to disable the "Show Answer" button as well.~~ **EDIT**: It is OK for the "Show Answer" button to be enabled if the last remaining attempt was correct because this matches the behavior of CAPA problems (which is what we want). Just keeping this note here in case the question comes up again.

* There is a possible regression regarding feedback messages: When misplacing an item in the final attempt, the feedback message says "Misplaced item was returned to item bank.". This does not reflect reality; if there are no attempts left, all items stay on the board, even if they were not placed correctly.

arbrandes commented:
@itsjeyd, thanks once again for your detailed, and very pertinent, review.  I've addressed the following in 8a260ed:

* The "Show Answer" button now cycles between "show mode" and "hide mode", which gives a way to return to the final submission, exactly as it was when first drawn.  This moots the need to disable the button once the correct answers are drawn.
* The failing keyboard help integration test could be reliably reproduced, and was fixed.
* I adopted your comment suggestion for the `isinstance(zone_title, dict)` verbatim.
* Dummy translations provided as requested.

However, the following represent behavior currently on master:

* When submitting the final attempt, misplaced items are left on the board.  This is contrary both to what the feedback messages say, and to what happens on previous attempts.
* When refreshing the page after the final submission, misplaced items are *not* left on the board.  This matches the behavior when submitting attempts prior to the last one.

My question regarding the above is: is it a bug, or is the behavior as intended?  If it's a bug, we should probably handle it on a separate issue - or maybe just a separate commit.  Thoughts?

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -150,6 +160,86 @@ def test_max_attempts_reached_submit_and_reset_disabled(self):
+        for item in self._get_items_with_zone(self.items_map).values():
+            zone_titles = [zones[zone_id] for zone_id in item.zone_ids]
+            # When showing answers, correct items are placed as if assessment_mode=False
+            self.assert_placed_item(item.item_id, zone_titles, assessment_mode=False)
```
@arbrandes This block is duplicated below. Factor it out into a separate `_assert_show_answer_item_placement` method, perhaps? When you do, keep the inline comment from L182 please, it's very useful :)

It might make sense to include the `for` loop below as well. That way, both tests would be checking the placement of decoy items.

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -150,6 +160,86 @@ def test_max_attempts_reached_submit_and_reset_disabled(self):
+
+        for item in self._get_items_without_zone(self.items_map).values():
+            self.assertNotDraggable(item.item_id)
+            element = self._get_item_by_value(item.item_id)
```
@arbrandes Nit: `element` is a bit generic. You could rename the loop variable to `item_definition`; that would make it possible to rename `element` to `item`.

itsjeyd commented on tests/integration/test_interaction_assessment.py:
```diff
@@ -150,6 +160,86 @@ def test_max_attempts_reached_submit_and_reset_disabled(self):
+
+        # Precondition: check item placement
+        item_description = self._get_placed_item_by_value(0).find_element_by_css_selector('.sr').text
+        self.assertEqual(item_description, 'Placed in: {}'.format(TOP_ZONE_TITLE))
```
@arbrandes This logic is reused twice below. You could put it into a nested `check_item_placement` function to DRY out this test.

itsjeyd commented on tests/unit/test_advanced.py:
```diff
@@ -402,6 +416,43 @@ def test_get_user_state_does_not_include_correctness(self):
+        (3, 3, False),
+    )
+    @ddt.unpack
+    def test_show_answer_validation(self, max_attempts, attempts, expect_validation_error):
```
@arbrandes Please add a docstring to this test and `test_get_correct_state` below.

itsjeyd commented:
@arbrandes Thank you for the additional changes and detailed reply to my comments. I really like how it's possible to toggle between correct and actual answers now. This matches the behavior for CAPA problems even more closely -- nice work!

Regarding your questions about intended behavior:

> * When submitting the final attempt, misplaced items are left on the board. ...

This is intended behavior: After submitting the final attempt the learner can not interact with the problem anymore, so we "freeze" their solution.

> ... This is contrary both to what the feedback messages say, and to what happens on previous attempts.

The behavior that you get for non-final attempts (misplaced items are returned to the item bank) is supposed to support learners in finding the correct solution:

* By leaving correctly placed items on the board, we enable them to build on a partial solution (instead of having to start from scratch).

* By returning misplaced items to the item bank, we clearly indicate which items a learner needs to focus on next. This (hopefully) prevents the process of completing the DnDv2 problem from turning into a guessing game: If all items stayed on the board, it might be very hard for the learner to tell misplaced items and correctly-placed items apart.

So, to make things consistent, we'll just need to fix the feedback message that comes up after the final attempt (it should not say "Misplaced item/s was/were returned to item bank."). If you could do that as part of this PR, that would be great. It looks like something that should be fairly easy to implement. If that's not the case, let me know and we can see about handling it in a separate PR.

> * When refreshing the page after the final submission, misplaced items are not left on the board. This matches the behavior when submitting attempts prior to the last one.

This seems like a bug to me, but then again I might have missed some of the more recent discussions we had with upstream about DnDv2.1 functionality. I will check with @bradenmacdonald and get back to you. (Additional note: Since the behavior is present on master, changing it is not necessarily part of the scope of this task. However, we are trying to wrap up DnDv2.1, and it wouldn't hurt if we could change the behavior as part of this PR; if it doesn't blow up the scope to much.)

---

Please use separate commits for any additional changes you make; that makes it easier for me to see what's new. You can squash commits again once I give a final thumbs-up. Thanks!

arbrandes commented:
@itsjeyd, I've addressed your comments, and a few quality issues, in separate commits headed by fb809c1.  The sandbox is being updated, and should be ready by the time you read this.

I have yet to address the "misplaced items are removed from the board after a refresh" thing, though it seems like a bug to me as well.  I suspect it won't be a particularly easy fix, but I'm going to do some digging and report back.

itsjeyd commented on drag_and_drop_v2/translations/eo/LC_MESSAGES/text.po:
```diff
@@ -634,11 +634,17 @@ msgstr[0] "Çörréçtlý pläçéd {correct_count} ïtém. Ⱡ'σяєм ιρѕ
 msgstr[0] "Mïspläçéd {misplaced_count} ïtém. Ⱡ'σяєм ιρѕυм ∂σłσя ѕιт αмєт,#"
 msgstr[1] "Mïspläçéd {misplaced_count} ïtéms. Ⱡ'σяєм ιρѕυм ∂σłσя ѕιт αмєт, #"
 
+#: utils.py:32
```
@arbrandes Is this string really defined in the same line as the previous one?

itsjeyd commented:
@arbrandes Thanks for fixing the "Misplaced n items" message, and for cleaning up the tests. This is looking good!

> I have yet to address the "misplaced items are removed from the board after a refresh" thing, though it seems like a bug to me as well. I suspect it won't be a particularly easy fix, but I'm going to do some digging and report back.

I talked to @bradenmacdonald about this, and he also thinks it's a bug. We'd like to see it addressed in the context of this PR, unless it blows up the scope too much. Let me know what you find -- if the issue turns out to be difficult to address, we can open up a new ticket for it.

itsjeyd commented:
@arbrandes Any updates regarding the "misplaced items are removed from the board after a refresh" issue? :)

itsjeyd commented:
@arbrandes To avoid delaying this too much, let's handle the "misplaced items are removed from the board after a refresh" issue separately :+1: 

I created https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/101 for upstream review. Some additional changes recently landed upstream; I updated `open-craft:master` to include them. Could you please rebase your branch?

CC @bradenmacdonald

arbrandes commented:
Hi @itsjeyd, apologies for the delay.  I've addressed your last line comment, rebased, and the sandbox should be updated by the time you read this.

I'm still looking into that last issue, but I agree that it would be better to handle it separately.

itsjeyd commented:
@arbrandes Welcome back, and thanks for the updates :) The feature is working as expected on the sandbox.

I noticed a small layout issue. The border between the "Reset" and "Show Answer" buttons is longer than the border between the "Keyboard Help" and "Reset" buttons:

![button-borders](https://cloud.githubusercontent.com/assets/961441/18466882/e06cf194-799d-11e6-9e1f-93bcbbe2ffd9.png)

I initially thought stale assets might be causing it, but updating static assets and restarting the sandbox via

```sh
paver update_assets --settings=aws
/edx/bin/supervisorctl restart edxapp:*
```

did not help.

Please fix this issue and address the failing build (looks like it's just a duplicate import). I'll add an explicit thumbs-up to https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/101 and ping the upstream reviewers once that's done.

arbrandes commented:
@itsjeyd, whoops!  Yes, that was a double import I missed during the last rebase.  Fixed.  I also fixed the separator heights.  Sandbox is being rebuilt.
